x-pack/plugins/fleet/server/services/agents/unenroll_action_runner.ts

```diff
@@ -48,23 +54,19 @@ export async function unenrollBatch(
   },
   skipSuccess?: boolean
 ): Promise<{ items: BulkActionResult[] }> {
-  // Filter to those not already unenrolled, or unenrolling
-  const agentsEnrolled = givenAgents.filter((agent) => {
-    if (options.revoke) {
-      return !agent.unenrolled_at;
-    }
-    return !agent.unenrollment_started_at && !agent.unenrolled_at;
-  });
-
-  const hostedPolicies = await getHostedPolicies(soClient, agentsEnrolled);
-
+  const hostedPolicies = await getHostedPolicies(soClient, givenAgents);
   const outgoingErrors: Record<Agent['id'], Error> = {};
 
   // And which are allowed to unenroll
   const agentsToUpdate = options.force
-    ? agentsEnrolled
-    : agentsEnrolled.reduce<Agent[]>((agents, agent) => {
-        if (isHostedAgent(hostedPolicies, agent)) {
+    ? givenAgents
+    : givenAgents.reduce<Agent[]>((agents, agent) => {
+        if (
+          (options.revoke && agent.unenrolled_at) ||
+          (!options.revoke && (agent.unenrollment_started_at || agent.unenrolled_at))
+        ) {
+          outgoingErrors[agent.id] = new FleetError(`Agent ${agent.id} already unenrolled`);
```

Comment:
Changed the implementation of unenroll to be consistent with reassign: if an agent is already unenrolled, the action is set to failed for the agent.
Previously the unenrolled agents were filtered out, so the action didn't include them. This is problematic in agent activity, as those agents don't have any trace of being actioned.